llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp

```diff
@@ -953,6 +933,68 @@ static void updateGenXMDNodes(llvm::Function *F, genx::KernelMDOp MD,
   }
 }
 
+// A functor which updates ESIMD kernel's uint64_t metadata in case it is less
+// than the given one. Used in callgraph traversal to update nbarriers or SLM
+// size metadata. Update is performed by the '()' operator and happens only
+// when given function matches one of the kernels - thus, only reachable kernels
+// are updated.
+// TODO: 1) In general this function is supposed to handle intrinsics
+// translated into kernel's metadata. So, the primary/intended usage model is
+// when such intrinsics are called from kernels.
+// 2) For now such intrinsics are also handled in functions directly called
+// from kernels and being translate into those caller-kernel meeven though such
+// behaviour is not fully specified/documented.
+// 3) This code (or the code in FE) must verify that slm_init or other such
+// intrinsic is not called from another module because kernels in that other
+// module would not get updated meta data attributes.
+struct UpdateUint64MetaDataToMaxValue {
```

Comment:
This refactors the original `updateGenXMDNodes`:
1) Call graph traversal is factored out into `traverseCallgraphUp` above. This functor represents call graph action.
2) It is slightly optimized to pre-select candidate nodes for fewer actions in the node action function.